src/cpp/src/tokenizer.cpp

```diff
@@ -117,184 +114,108 @@ class Tokenizer::TokenizerImpl {
-            if (val.empty()) { read_json_param(data, key_name, val); }
-        };
-        read_token_str(pad_token_key_name, m_pad_token);
-        read_token_str(bos_token_key_name, m_bos_token);
-        read_token_str(eos_token_key_name, m_eos_token);
-
-        // if special tokens are not loaded directly, try to read
-        // if they are in the format {"bos_token": { "content": "<s>",... }}
-        auto read_token_content_str = [&data](std::string key_name, std::string& val) {
-            if (val.empty() && data.contains(key_name)) { read_json_param(data[key_name], "content", val); }
-        };
-        read_token_content_str(pad_token_key_name, m_pad_token);
-        read_token_content_str(bos_token_key_name, m_bos_token);
-        read_token_content_str(eos_token_key_name, m_eos_token);
-
-        // if pad_token not found use eos_token as pad_token
-        if (m_pad_token.empty() && !m_eos_token.empty()) {
-            m_pad_token = m_eos_token;
-        }
-
-        // special token ids integer representation are already defined
-        if (m_pad_token_id != -1 && m_bos_token_id != -1 && m_eos_token_id != -1)
-            return ;
-
-        // values are stored as {"added_tokens_decoder": {"0": {"content": "<pad>"}}}
-        // token id is a key in the form of a string, need to do std::stoi
-        std::string spec_tokens_key_name = "added_tokens_decoder";
-        if (!data.contains(spec_tokens_key_name))
-            return ;
-
-        // if added_tokens_decoder has different format items() will not fail
-        for (auto& [key, value] : data[spec_tokens_key_name].items()) {
-            if (!value.contains("content"))
-                continue;
-            auto content = value["content"];
-            if (m_pad_token_id == -1 && content == m_pad_token)
-                m_pad_token_id = std::stoi(key);
-            if (m_bos_token_id == -1 && content == m_bos_token)
-                m_bos_token_id = std::stoi(key);
-            if (m_eos_token_id == -1 && content == m_eos_token)
-                m_eos_token_id = std::stoi(key);
-        }
-
-        // if pad_token_id not found use eos_token_id as pad_token_id
-        // todo: read m_pad_token_id from tokenizer rt_info once implemented in tokenizers (CVS-144174)
-        if (m_pad_token_id == -1 && m_eos_token_id != -1) {
-            m_pad_token_id = m_eos_token_id;
-        }
+    TokenizerImpl(
+        std::string& tokenizer_model_str,
+        ov::Tensor& tokenizer_weights_tensor,
+        std::string& detokenizer_model_str,
+        ov::Tensor& detokenizer_weights_tensor,
+        const ov::AnyMap& properties
+    ) {
+        auto core = *get_core();
+
+        auto ov_tokenizer = core.read_model(tokenizer_model_str, tokenizer_weights_tensor);
+        auto ov_detokenize = core.read_model(detokenizer_model_str, detokenizer_weights_tensor);
+        *this = TokenizerImpl(std::make_pair(ov_tokenizer, ov_detokenize), properties);
```

Comment:
instead of such hacks, we can check whether tokenizer_model_str and tokenizer_weights_tensor are not empty, then call read_model. 

In this case we don' need `TokenizerImpl(const std::pair<std::shared_ptr<ov::Model>, std::shared_ptr<ov::Model>>& models,  const ov::AnyMap& properties)`

Reply:
I removed that hach with `*this = ...`, now we have only one `TokenizerImpl` ctor, and it's `TokenizerImpl(const std::pair<std::shared_ptr<ov::Model>, std::shared_ptr<ov::Model>>& models,  const ov::AnyMap& properties)`

I don't quite understand why did you want to remove it. It's very handy, i can use for all cases, whether it's a path or a buffer, i just read in the main `Tokenzier` class and call `TokenizerImpl` for `m_pimp` with shared_ptr ov::Model. 

Can you please take a look to the lates version and tell whether it's ok?